New Meetup Spice

alohaas opened the pull request:
Hey team,

I wanted to submit the work I've done toward creating a new IA for Meetups.
The original idea can be found here: https://duck.co/ideas/idea/1126/already-an-instant-answers-for-meetup
Please let me know if I've overlooked anything!

Thanks!

Olivia
# Spice Pull Request Template

**What does your Instant Answer do?**
The Meetup IA generates meetup groups for a given US city.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It can provide a user with a survey of the social or entrepreneurial culture of a place they call home, or a place he or she plans to visit. It bypasses creating a meetup account to browse local groups and can speed up the search for a distant group for an existing user.

**What is the data source for your Instant Answer? (Provide a link if possible)**
http://www.meetup.com/meetup_api/
I'm using meetup.com's API method /2/groups + a city/state/country query

**Why did you choose this data source?**
I chose this data source because it uses one API Key to authenticate and it provides a method for searching groups by city. 

**Are there any other alternative (better) data sources?**
Meetup.com's API is probably the best data source, however there are alternate API methods available which could provide alternative means of triggering meetup IAs.

**What are some example queries that trigger this Instant Answer?**
"meetups in Philadelphia", "meetups near me", "meetup groups"

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
https://duck.co/ideas/idea/1126/already-an-instant-answers-for-meetup

**Which existing Instant Answers will this one supersede/overlap with?**
Have not come across any.

**Are you having any problems? Do you need our help with anything?**
Perhaps! I've navigated a lot of initial problems, but my Perl may be a little off :)

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
API keys for Meetup.com are free. I built the IA using my personal meetup IA. 

**Where did you hear about DuckDuckHack? (For first time contributors)**
Kind of a long story, but ultimately it was from the Duck Duck Go team. I was directed to the team and ultimately the community though my good friend Corinne at Girl Develop It and my friends David and Adam who are active in the Philly tech scene.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
<img width="1439" alt="screenshot 2015-07-22 23 48 40" src="https://cloud.githubusercontent.com/assets/4480279/8842768/601a3daa-30cc-11e5-927e-13ab818e259a.png">
<img width="1439" alt="screenshot 2015-07-22 23 48 47" src="https://cloud.githubusercontent.com/assets/4480279/8842769/63064afe-30cc-11e5-99bf-98fcc665f508.png">
## Checklist

Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information
[X] Wrote test file and added to t/ directory
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 8

    - Mac OSX
        [X] Google Chrome
        [X] Firefox
        [] Opera
        [X] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

https://duck.co/ia/view/meetup

moollaza commented:
@alohaas thanks for your contribution!

Can you please fill out the [pull request template](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/pull_request_template_spice.md) and edit your initial comment. Once completed, we'll be sure to review your PR and give you some feedback.

Thanks!

alohaas commented:
@moollaza sure thing. Apologies for the oversight!

alohaas commented:
# Spice Pull Request Template

**What does your Instant Answer do?**
The Meetup IA generates meetup groups for a given US city.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It can provide a user with a survey of the social or entrepreneurial culture of a place they call home, or a place he or she plans to visit. It bypasses creating a meetup account to browse local groups and can speed up the search for a distant group for an existing user.

**What is the data source for your Instant Answer? (Provide a link if possible)**
http://www.meetup.com/meetup_api/
I'm using meetup.com's API method /2/groups + a city/state/country query

**Why did you choose this data source?**
I chose this data source because it uses one API Key to authenticate and it provides a method for searching groups by city. 

**Are there any other alternative (better) data sources?**
Meetup.com's API is probably the best data source, however there are alternate API methods available which could provide alternative means of triggering meetup IAs.

**What are some example queries that trigger this Instant Answer?**
"meetups in Philadelphia", "meetups near me", "meetup groups"

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**


**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
https://duck.co/ideas/idea/1126/already-an-instant-answers-for-meetup

**Which existing Instant Answers will this one supersede/overlap with?**
Have not come across any.

**Are you having any problems? Do you need our help with anything?**
Perhaps! I've navigated a lot of initial problems, but my Perl may be a little off :)

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
API keys for Meetup.com are free. I built the IA using my personal meetup IA. 

**Where did you hear about DuckDuckHack? (For first time contributors)**
Kind of a long story, but ultimately it was from the Duck Duck Go team. I was directed to the team and ultimately the community though my good friend Corinne at Girl Develop It and my friends David and Adam who are active in the Philly tech scene.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
<img width="1439" alt="screenshot 2015-07-22 23 48 40" src="https://cloud.githubusercontent.com/assets/4480279/8842768/601a3daa-30cc-11e5-927e-13ab818e259a.png">
<img width="1439" alt="screenshot 2015-07-22 23 48 47" src="https://cloud.githubusercontent.com/assets/4480279/8842769/63064afe-30cc-11e5-99bf-98fcc665f508.png">

## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information
[X] Wrote test file and added to t/ directory
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 8

    - Mac OSX
        [X] Google Chrome
        [X] Firefox
        [] Opera
        [X] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

moollaza commented:
@alohaas whoops, did you mean to close this?

alohaas commented:
Hey @moollaza I did not! I accidentally posted the template twice and wanted to delete one of the duplicate comments. I've been trying to reopen it, but can't figure it out :)

alohaas commented:
@moollaza I can create a new pull request if needed. I haven't found any documentation on how to reopen it...

moollaza commented:
>I did not! I accidentally posted the template twice and wanted to delete one of the duplicate comments. I've been trying to reopen it, but can't figure it out :)
@alohaas aah no problem. I have the power to reopen them :+1:

jagtalon commented:
Just made an IA Page for this https://duck.co/ia/view/meetup

jagtalon commented:
@alohaas 

Please add this to `Spice.add` (inside the `templates` object):

```javascript
variants: {
    tileSnippet: "large"
}
```

So that we'll get this (consistent tile sizes):

![screen shot 2015-09-17 at 2 54 04 pm](https://cloud.githubusercontent.com/assets/81969/9942929/fb9c5f1e-5d4b-11e5-82b1-9cf137f8072c.png)

Instead of this:

![screen shot 2015-09-17 at 2 51 58 pm](https://cloud.githubusercontent.com/assets/81969/9942933/02e04fce-5d4c-11e5-9b7c-d41cf956a0cc.png)

jagtalon commented on share/spice/meetup/meetup.js:
```diff
@@ -0,0 +1,45 @@
+              count: api_result.results.length,
+              itemType: 'meetups',
+              sourceName: "Meetup.com",
+              sourceUrl: 'http://www.meetup.com/'
```
Can we make the "More at" link to what the user is looking for like http://www.meetup.com/find/events/?allMeetups=true&radius=10&userFreeform=phi&mcId=c19019&change=yes for Philly?

alohaas replied:
For sure! Great idea.

jagtalon commented on share/spice/meetup/meetup.js:
```diff
@@ -0,0 +1,45 @@
+          meta: {
+              count: api_result.results.length,
+              itemType: 'meetups',
+              sourceName: "Meetup.com",
```
We usually drop the last part of the `sourceName` (the technical name is escaping me right now but it's the ".com" part) like in https://duckduckgo.com/?q=ray+bradbury+books&ia=products we just say "Amazon" instead of "Amazon.com". So here, it would be just "Meetup. Ah, it's TLD. Just remembered.

alohaas replied:
Gotcha. I can make this change.

jagtalon commented:
@alohaas we usually cache for 30d https://github.com/duckduckgo/zeroclickinfo-spice/search?utf8=%E2%9C%93&q=30d on IAs that we know that won't update that often. Do you think that's reasonable or is that too long?

jagtalon commented on share/spice/meetup/meetup.js:
```diff
@@ -0,0 +1,45 @@
+    "use strict";
+    env.ddg_spice_meetup = function(api_result){
+
+        if (!api_result || api_result.error || !api_result.results) {
```
Looks like it should be `errors` instead of just `error` that we should watch out for http://www.meetup.com/meetup_api/docs/:

![screen shot 2015-09-17 at 3 13 05 pm](https://cloud.githubusercontent.com/assets/81969/9943320/a86de63e-5d4e-11e5-8078-25907b093081.png)

jagtalon replied:
We should probably exit early as well if `results.length` is 0

alohaas replied:
Thanks for catching this! I'll make the change.

jagtalon commented on t/Meetup.t:
```diff
@@ -0,0 +1,37 @@
+    ),
+
+    'meet group' => undef,
+    'meetups about cats' => undef,
```
:crying_cat_face:

alohaas replied:
Can change this to reference 'meetups about dogs.'

jagtalon commented on lib/DDG/Spice/Meetup.pm:
```diff
@@ -0,0 +1,62 @@
+            twitter => "livhaas";
+
+spice from => '([^/]+)/?(?:([^/]+)/?(?:([^/]+)|)|)';
+spice to => 'https://api.meetup.com/2/groups?&key={{ENV{DDG_SPICE_MEETUP_APIKEY}}}&country=$2&city=$1&state=$3&radius=50.0&page=20&order=members&callback={{callback}}';
```
It looks like it makes the wrong request when I'm searching outside of my city. So for example I want to see the meetups in San Francisco, I type in "meetups in san francisco", but it makes a request to this endpoint:

`http://window-roman.codio.io:5000/js/spice/meetup/san%20francisco/US/PA`

So it sends the right city (San Francisco), but it doesn't send the right state (PA).

jagtalon replied:
I guess since we don't have our geocoder exposed to the Spices (the thing that makes sense of the string "San francisco") I guess we have two options here:

- Only trigger on "meetups near me" and the like
- See if Meetup.com API can accept a free form string such as "San Francisco" without having the need to specify the state and the country (which is how https://duckduckgo.com/?q=weather+in+los+angeles&ia=weather works)

daxtheduck commented:
Automated screenshot from https://beta.duckduckgo.com
                ![](https://ia-screenshots.s3.amazonaws.com/meetup_index.png)
                mobile: https://ia-screenshots.s3.amazonaws.com/meetup_mobile.png

daxtheduck commented:
Automated screenshot fromhttps://beta.duckduckgo.com/?q=meetup
                ![](https://ia-screenshots.s3.amazonaws.com/meetup_index.png)
                mobile: https://ia-screenshots.s3.amazonaws.com/meetup_mobile.png

daxtheduck commented:
Automated screenshot fromhttps://beta.duckduckgo.com/?q=meetup
                ![](https://ia-screenshots.s3.amazonaws.com/meetup_index.png)
                mobile: https://ia-screenshots.s3.amazonaws.com/meetup_mobile.png

daxtheduck commented:
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=meetup)
                ![](https://ia-screenshots.s3.amazonaws.com/meetup_index.png)
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/meetup_mobile.png)

daxtheduck commented:
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=meetup)
                ![](https://ia-screenshots.s3.amazonaws.com/meetup_index.png)
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/meetup_mobile.png)

daxtheduck commented:
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=meetup)
                ![](https://ia-screenshots.s3.amazonaws.com/meetup_index.png)
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/meetup_mobile.png)

daxtheduck commented:
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=meetup)
                ![](https://ia-screenshots.s3.amazonaws.com/meetup_index.png)
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/meetup_mobile.png)

daxtheduck commented:
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=meetup)
                ![](https://ia-screenshots.s3.amazonaws.com/meetup_index.png)
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/meetup_mobile.png)

philoniare commented:
This is great! I was also thinking of doing a meetup IA before discovering this. Is there anything I can contribute to move this into production? Also, instead of sorting by number of people, maybe a more common use case is to sort by date, so that upcoming events show up.

alohaas commented:
Hey @philoniare, I like the idea of sorting by date of upcoming event. It's been a while since I worked with the API, but I seem to remember it having an array of all upcoming events. Maybe we could show them in the detail view. 
I'll need to touch base with @moollaza to see what else needs to be changed before it can go live. We may just need to use a different API request and/or think about other triggering methods.

philoniare commented:
Hi @alohaas, thanks for the response. I'll also look into what I can do with it and learn some perl :panda_face:

daylenca commented:
The instant answer is not working in Microsoft Edge (link - https://beta.duckduckgo.com/?q=meetup&ia=meetup). Is this the same in other browsers?

tagawa commented:
This doesn't seem to be appearing for me in other browsers (Firefox, Vivaldi, Safari) either.

daylenca commented:
@alohaas The instant answer isn't triggering

pjhampton commented:
Hey @alohaas 

[DuckDuckHack is now in Maintenance Mode](https://duckduckhack.com/) and from now on, we are only accepting issues and PRs for essential bugs and bug fixes.

Unfortunately, this is non-essential and will be closed.

We appreciate you taking the time to contribute and apologize for not being able to include your work.
